### ISSUE TITLE
Update Linkedin to support api/v2

### ISSUE THREAD

**Issue by kchowanski** [aspnet-contrib/AspNet.Security.OAuth.Providers#277]
Hello,
Linkedin enforce users to migrate to API version 2.0 by March 1.

"...all developers need to migrate to Version 2.0 of our APIs and OAuth 2.0 by March 1."

https://engineering.linkedin.com/blog/2018/12/developer-program-updates.

**Comment by Mattzard** [comment]
Whilst this work is being completed, I don't know if it would be recommended to also make the call to return the email address within the claims? 

For version 2 of the API email address has been separated into a different call (_https://api.linkedin.com/v2/emailAddress?q=members&projection=(elements*(handle~))_) and isn't returned with the new _lite_profile_, yet previously it would have been returned in the claims for version 1 of the API so may aid people who are using that already.

**Comment by martincostello** [comment]
This sounds similar to how the GitHub and Bitbucket providers work, where emails need to be requested separately with a different HTTP request.

**Comment by jandieg** [comment]
Hi, is this related? Just using a new LinkedIn api (v2)

> An unhandled exception occurred while processing the request.
> HttpRequestException: An error occurred while retrieving the user profile.
> AspNet.Security.OAuth.LinkedIn.LinkedInAuthenticationHandler.CreateTicketAsync(ClaimsIdentity identity, AuthenticationProperties properties, OAuthTokenResponse tokens)
> 
> Exception: An error was encountered while handling the remote login.
> Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler<TOptions>.HandleRequestAsync()

**Comment by odahcam** [comment]
I'm having some trouble to use the LinkeIn authentication as well. In my case, when the user comes back from the LinkedIn's authorization page to my route callback, the first thing I'm doing is:

```csharp
var loginInfo = await AuthenticationManager.GetExternalLoginInfoAsync();
```

But this method call, returns `null` for LinkedIn while works perfectly with Twitter, Google SignIn, Facebook and Instagram.

Are you here with the same problem?

**Comment by jandieg** [comment]
I think it is part of the same. Current implementations assume the **email** is coming as a claim, but on Linkedin API v2 it is no longer.

@devs is there a workaround for this? At least to ignore the email claim?

**Comment by odahcam** [comment]
As I could see `loginInfo` is completely empty and `nameidentifier` is `0`. 😕

**Comment by Mattzard** [comment]
> I think it is part of the same. Current implementations assume the **email** is coming as a claim, but on Linkedin API v2 it is no longer.
> 
> @devs is there a workaround for this? At least to ignore the email claim?

I was able to get this working by downloading the code for LinkedIn (here: https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/tree/dev/src/AspNet.Security.OAuth.LinkedIn), importing this as a project in my solution, and then configuring the LinkedInAuthenticationHandler class to make an additional call to LinkedIn's API to return the Email Address. 

I believe this is the code snippet I have used:

            var emailAddressRequest = new HttpRequestMessage(HttpMethod.Get, "https://api.linkedin.com/v2/emailAddress?q=members&projection=(elements*(handle~))");
            emailAddressRequest.Headers.Add("x-li-format", "json");
            emailAddressRequest.Headers.Authorization = new AuthenticationHeaderValue("Bearer", tokens.AccessToken);

            var emailAddressResponse = await Backchannel.SendAsync(emailAddressRequest, Context.RequestAborted);
            if (!emailAddressResponse.IsSuccessStatusCode)
            {
                Logger.LogError("An error occurred while retrieving the email address: the remote server " +
                "returned a {Status} response with the following payload: {Headers} {Body}.",
                /* Status: */ emailAddressResponse.StatusCode,
                /* Headers: */ emailAddressResponse.Headers.ToString(),
                /* Body: */ await emailAddressResponse.Content.ReadAsStringAsync());

                throw new HttpRequestException("An error occurred while retrieving the email address.");
            }
            var emailAddressPayload = JObject.Parse(await emailAddressResponse.Content.ReadAsStringAsync());

            string emailAddress = emailAddressPayload.SelectToken("elements[0].handle~.emailAddress").ToString();

I don't really recommend following my steps as this prevents me from easily obtaining updates to the LinkedIn library if any become available.

**Comment by odahcam** [comment]
It would be nice to see a official fix here. I don't intend to download the source and change it. That's a bad practice and can generate problems I have no time to deal with. 🙁

**Comment by jandieg** [comment]
Maybe Mattzard addition can be the solution?

Btw, all new LinkedIn apps are being forced to use the latest API which does not work. So this is becoming like a critical issue.

**Comment by martincostello** [comment]
The fastest way to get a fix would be if any affected parties were to raise a pull request with a proposed solution and we can go from there, particularly for a fix for the current ASP.NET Core 2.x providers.

Otherwise this is something that I can pick up as part of changes to add support for ASP.NET Core 3.0 that I've started working on. I'm adding tests to this repo as part of those changes, and it's uncovered a handful of issues in the existing providers such as #284.

**Comment by jdealarcon** [comment]
Thanks @Mattzard  your changes works perfect!!
Here are code mixing both endpoint
**LinkedInAuthenticationHandler.cs**
`        protected override async Task<AuthenticationTicket> CreateTicketAsync([NotNull] ClaimsIdentity identity,
            [NotNull] AuthenticationProperties properties, [NotNull] OAuthTokenResponse tokens)
        {


            //------@Mattzard Code ------------------------------------------
            var emailAddressRequest = new HttpRequestMessage(HttpMethod.Get, "https://api.linkedin.com/v2/emailAddress?q=members&projection=(elements*(handle~))");
            emailAddressRequest.Headers.Add("x-li-format", "json");
            emailAddressRequest.Headers.Authorization = new AuthenticationHeaderValue("Bearer", tokens.AccessToken);

            var emailAddressResponse = await Backchannel.SendAsync(emailAddressRequest, Context.RequestAborted);
            if (!emailAddressResponse.IsSuccessStatusCode)
            {
                Logger.LogError("An error occurred while retrieving the email address: the remote server " +
                "returned a {Status} response with the following payload: {Headers} {Body}.",
                /* Status: */ emailAddressResponse.StatusCode,
                /* Headers: */ emailAddressResponse.Headers.ToString(),
                /* Body: */ await emailAddressResponse.Content.ReadAsStringAsync());

                throw new HttpRequestException("An error occurred while retrieving the email address.");
            }
            var emailAddressPayload = JObject.Parse(await emailAddressResponse.Content.ReadAsStringAsync());

            string emailAddress = emailAddressPayload.SelectToken("elements[0].handle~.emailAddress").ToString();
            //-------end @Mattzard----------------------------------------
            var address = Options.UserInformationEndpoint;

            // If at least one field is specified,
            // append the fields to the endpoint URL.
            //if (Options.Fields.Count != 0)
            //{
            //    address = address.Insert(address.LastIndexOf("~") + 1, $":({ string.Join(",", Options.Fields)})");
            //}
           // address = "https://api.linkedin.com/v2/me";

            var request = new HttpRequestMessage(HttpMethod.Get, address);
            request.Headers.Add("x-li-format", "json");
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", tokens.AccessToken);

            var response = await Backchannel.SendAsync(request, Context.RequestAborted);
            if (!response.IsSuccessStatusCode)
            {
                Logger.LogError("An error occurred while retrieving the user profile: the remote server " +
                                "returned a {Status} response with the following payload: {Headers} {Body}.",
                                /* Status: */ response.StatusCode,
                                /* Headers: */ response.Headers.ToString(),
                                /* Body: */ await response.Content.ReadAsStringAsync());

                throw new HttpRequestException("An error occurred while retrieving the user profile.");
            }
            
            var payload = JObject.Parse(await response.Content.ReadAsStringAsync());
            string nombre = payload.SelectToken("localizedFirstName").ToString();
            string apellidos = payload.SelectToken("localizedLastName").ToString();
            string nombreCompleto = string.Format("{0} {1}", nombre, apellidos);
            var nombreCompletoProperty = new JProperty("nombreCompleto", nombreCompleto);
            payload.Last.AddAfterSelf(nombreCompletoProperty);
            var emailProperty = new JProperty("emailAddress", emailAddress);
            payload.Last.AddAfterSelf(emailProperty);
          

            var principal = new ClaimsPrincipal(identity);

            
            var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, payload);
            context.RunClaimActions(payload);

            await Options.Events.CreatingTicket(context);
            return new AuthenticationTicket(context.Principal, context.Properties, Scheme.Name);
        }`

**LinkedInAuthenticationOptions.cs**
` public class LinkedInAuthenticationOptions : OAuthOptions
    {
        public LinkedInAuthenticationOptions()
        {
            ClaimsIssuer = LinkedInAuthenticationDefaults.Issuer;

            CallbackPath = new PathString(LinkedInAuthenticationDefaults.CallbackPath);

            AuthorizationEndpoint = LinkedInAuthenticationDefaults.AuthorizationEndpoint;
            TokenEndpoint = LinkedInAuthenticationDefaults.TokenEndpoint;
            UserInformationEndpoint = LinkedInAuthenticationDefaults.UserInformationEndpoint;


            Scope.Add("r_liteprofile");
            Scope.Add("r_emailaddress");

            ClaimActions.MapJsonKey(ClaimTypes.NameIdentifier, "id");
            ClaimActions.MapJsonKey(ClaimTypes.Email, "emailAddress");
            ClaimActions.MapJsonKey(ClaimTypes.Name, "nombreCompleto");
            ClaimActions.MapJsonKey(ClaimTypes.GivenName, "localizedFirstName");
            ClaimActions.MapJsonKey(ClaimTypes.Surname, "localizedLastName");
            //ClaimActions.MapJsonKey(Claims.MaidenName, "maidenName");
            //ClaimActions.MapJsonKey(Claims.ProfileUrl, "publicProfileUrl");
            //ClaimActions.MapJsonKey(Claims.PictureUrl, "pictureUrl");
            //ClaimActions.MapJsonKey(Claims.Industry, "industry");
            //ClaimActions.MapJsonKey(Claims.Summary, "summary");
            //ClaimActions.MapJsonKey(Claims.Headline, "headline");
            //ClaimActions.MapCustomJson(Claims.Positions, user => user["positions"]?.ToString());
            //ClaimActions.MapJsonKey(Claims.PhoneticFirstName, "phoneticFirstName");
            //ClaimActions.MapJsonKey(Claims.PhoneticLastName, "phoneticLastName");
            //ClaimActions.MapJsonKey(Claims.FormattedPhoneticName, "formattedPhoneticName");
            //ClaimActions.MapCustomJson(Claims.Location, user => user["location"]?.ToString());
            //ClaimActions.MapJsonKey(Claims.Specialties, "specialties");
            //ClaimActions.MapJsonKey(Claims.NumConnections, "numConnections");
            //ClaimActions.MapJsonKey(Claims.NumConnectionsCapped, "numConnectionsCapped");
            //ClaimActions.MapJsonKey(Claims.CurrentShare, "currentShare");
            //ClaimActions.MapCustomJson(Claims.PictureUrls, user => user["pictureUrls"]?.ToString());
        }

        /// <summary>
        /// Gets the list of fields to retrieve from the user information endpoint.
        /// See https://docs.microsoft.com/es-es/linkedin/consumer/integrations/self-serve/sign-in-with-linkedin for more information.
        /// </summary>
        public ISet<string> Fields { get; } = new HashSet<string>
        {
        //id The member's unique identifier Person URN 
        //firstName Member's first name Text 
        //lastName Member's last name Text 
        //profilePicture Member's profile picture URL Text 

         "id",
         "firstName",
        "lastName",
        "profilePicture(displayImage~:playableStreams)"

        };
    }`

**Comment by maftieu** [comment]
Could this be delivered before milestone 3.0?
LinkedIn has now switched off V1 api endpoints. All calls results to "410 Gone"!

Or do I need to include source code of the provider in my app to get this working?!

**Comment by martincostello** [comment]
Fix will come in a few days once I have the time to merge #298, check the state of master and arrange release and publishing.

**Comment by martincostello** [comment]
A prerelease version containing the fix merged from #298 is now available from the MyGet feed: https://www.myget.org/feed/aspnet-contrib/package/nuget/AspNet.Security.OAuth.LinkedIn/2.1.0-preview-0749

Please try it out and feedback whether your issues are resolved.

**Comment by odahcam** [comment]
I could not test because:

> Install-Package : Could not install package 'AspNet.Security.OAuth.LinkedIn 2.1.0-preview-0749'. You are trying to install this package into a project that targets 
'.NETFramework,Version=v4.5.2', but the package does not contain any assembly references or content files that are compatible with that framework. For more information, contact the package 
author.

**Comment by maftieu** [comment]
> A prerelease version containing the fix merged from #298 is now available from the MyGet feed: https://www.myget.org/feed/aspnet-contrib/package/nuget/AspNet.Security.OAuth.LinkedIn/2.1.0-preview-0749
> 
> Please try it out and feedback whether your issues are resolved.

I tried, and it works. I only use e-mail address from the profile.
Thanks!

**Comment by martincostello** [comment]
@odahcam The 2.x releases only support `netstandard2.0`.

**Comment by martincostello** [comment]
Closing as resolved by #298. Will look to release a 2.1.0 package with the fix to NuGet.org sometime in the next week.

**Comment by jandieg** [comment]
> A prerelease version containing the fix merged from #298 is now available from the MyGet feed: https://www.myget.org/feed/aspnet-contrib/package/nuget/AspNet.Security.OAuth.LinkedIn/2.1.0-preview-0749
> 
> Please try it out and feedback whether your issues are resolved.

Thanks. This is my first time with MyGet, am I doing something wrong?

```
Install-Package : NU1102: Unable to find package AspNet.Security.OAuth.LinkedIn with version (>= 2.1.1-preview-0768)
  - Found 14 version(s) in nuget.org [ Nearest version: 2.0.1 ]
  - Found 0 version(s) in Package source
  - Found 0 version(s) in Microsoft Visual Studio Offline Packages
```

**Comment by martincostello** [comment]
@jandieg You need to add the MyGet feed to your `NuGet.config` file. See here for an example: https://github.com/martincostello/alexa-london-travel-site/pull/292/files
